### ISSUE TITLE
Broadcasting with single dimension is currently not implemented

### ISSUE THREAD

**Issue by patricksongzy** [patricksongzy/corgi#1]
**Describe the bug**
Broadcasting, such as in matrix multiplying a tensor with dimensions [2], with one with dimensions [2, 2] will not backpropagate as expected.

* This impacts `Dense` layers if inputs are not given with the expected dimensions.

**To Reproduce**
```Rust
let w = arr![arr![1.0, 2.0], arr![8.0, 2.0]].tracked();
let x = arr![5.0, 3.0];

let mut result = Array::matmul((&x, false), (&w, true));
result.backward(None);
assert_eq!(w.gradient().unwrap(), arr![arr![5.0, 3.0], arr![5.0, 3.0]]);
```
Will fail, and the gradient for `w` will be `[[5.0, 0.0], [5.0, 0.0]]`.

This also means when inputting a tensor `x = arr![x_1, ..., x_n]` to a Dense layer with `n` inputs, the gradient will be incorrect.

* Workaround: input tensors with the proper dimensions, that is `x = arr![[x_1, ..., x_n]]` instead.
* This affects training, but only on first-layer weights for models with inputs which need to be broadcasted.
* This does not affect models which are trained using batch gradient descent, since such models have the proper dimensions specified.

**Additional context**
Tests have been added, and regressions will be easier to detect, given the addition of numerical gradient checking as well.
